hwy/contrib/thread_pool/thread_pool.h

```diff
@@ -918,20 +918,38 @@ class WorkerAdapter {
 
   void SetEpoch(uint32_t epoch) { epoch_ = epoch; }
 
-  // Split into separate wait/barrier functions because `ThreadFunc` latches
-  // the config in between them.
-  template <class Spin, class Wait,
-            HWY_IF_SAME(decltype(Wait().Type()), WaitType)>
-  void operator()(const Spin& spin, const Wait& wait) const {
+ private:
+  template <class Spin, class Wait>
+  HWY_INLINE void CallImpl(hwy::SizeTag<1> /* second_param_type_tag */,
+                           const Spin& spin, const Wait& wait) const {
     wait.UntilWoken(worker_, spin, epoch_);
   }
-
-  template <class Spin, class Barrier,
-            HWY_IF_SAME(decltype(Barrier().Type()), BarrierType)>
-  void operator()(const Spin& spin, const Barrier& barrier) const {
+  template <class Spin, class Barrier>
+  HWY_INLINE void CallImpl(hwy::SizeTag<2> /* second_param_type_tag */,
+                           const Spin& spin, const Barrier& barrier) const {
     barrier.WorkerReached(worker_, spin, epoch_);
   }
 
+ public:
+  // Split into separate wait/barrier functions because `ThreadFunc` latches
+  // the config in between them.
+  template <class Spin, class Param2>
+  hwy::EnableIf<
+      hwy::IsSame<hwy::RemoveCvRef<decltype(hwy::RemoveCvRef<Param2>().Type())>,
+                  WaitType>() ||
+      hwy::IsSame<hwy::RemoveCvRef<decltype(hwy::RemoveCvRef<Param2>().Type())>,
+                  BarrierType>()>
+  operator()(const Spin& spin, const Param2& wait_or_barrier) const {
+    this->CallImpl(
+        hwy::SizeTag<
+            hwy::IsSame<
```

Comment:
To maybe reduce the number of lines, we could have a `constexpr int kType = hwy::IsSame<
                hwy::RemoveCvRef<decltype(hwy::RemoveCvRef<Param2>().Type())>,
                WaitType>() ? 1 : 2;` then pass `SizeTag<kType>`?